packages/omgidl-parser/src/IdlNodes/IdlNode.ts

```diff
@@ -0,0 +1,191 @@
+  }
+
+  protected getNode(scopePath: string[], name: string): IdlNode {
+    const maybeNode = resolveScopedOrLocalNodeReference({
+      usedIdentifier: name,
+      scopeOfUsage: scopePath,
+      definitionMap: this.map,
+    });
+    if (maybeNode == undefined) {
+      throw new Error(
+        `Could not find node ${name} in ${scopePath.join("::")} referenced by ${
+          this.scopedIdentifier
+        }`,
+      );
+    }
+    return maybeNode;
+  }
+
+  protected getConstantNode(identifier: string): ConstantIdlNode {
+    const maybeConstantNode = this.getNode(this.scopePath, identifier);
+    if (!(maybeConstantNode instanceof ConstantIdlNode)) {
+      throw new Error(`Expected ${this.name} to be a constant in ${this.scopedIdentifier}`);
+    }
+    return maybeConstantNode;
+  }
+}
+
+export class ConstantIdlNode extends IdlNode<ConstantASTNode> {
+  private needsResolution = false;
+  constructor(scopePath: string[], astNode: ConstantASTNode, idlMap: Map<string, IdlNode>) {
+    super(scopePath, astNode, idlMap);
+    if (!SIMPLE_TYPES.has(astNode.type)) {
+      this.needsResolution = true;
+    }
+  }
+
+  get type(): string {
+    if (this.needsResolution) {
+      return this.getReferencedEnumNode().type;
+    }
+    return this.astNode.type;
+  }
+
+  private referencedEnumNode?: EnumIdlNode = undefined;
+  private getReferencedEnumNode(): EnumIdlNode {
+    if (this.referencedEnumNode == undefined) {
+      const maybeEnumNode = this.getNode(this.scopePath, this.astNode.type);
+      if (!(maybeEnumNode instanceof EnumIdlNode)) {
+        throw new Error(`Expected ${this.astNode.type} to be an enum in ${this.scopedIdentifier}`);
+      }
+      this.referencedEnumNode = maybeEnumNode;
+    }
+    return this.referencedEnumNode;
+  }
+
+  get isConstant(): true {
+    return true;
+  }
+
+  get value(): ConstantValue {
+    if (typeof this.astNode.value === "object") {
+      return this.getConstantNode(this.astNode.value.name).value;
+    }
+    return this.astNode.value;
+  }
+
+  toIDLMessageDefinitionField(): IDLMessageDefinitionField {
+    return {
+      name: this.name,
+      type: normalizeType(this.type),
+      value: this.value,
+      isConstant: true,
+      isComplex: false,
+      ...(this.astNode.valueText != undefined ? { valueText: this.astNode.valueText } : undefined),
+    };
+  }
+}
+
+export class EnumIdlNode extends IdlNode<EnumASTNode> {
+  constructor(scopePath: string[], astNode: EnumASTNode, idlMap: Map<string, IdlNode>) {
+    super(scopePath, astNode, idlMap);
+  }
+
+  get type(): string {
+    return "uint32";
+  }
+
+  get isComplex(): boolean {
+    return false;
+  }
+  get arrayLengths(): number[] | undefined {
+    return undefined;
+  }
+
+  get arrayUpperBound(): number | undefined {
+    return undefined;
+  }
+
+  get upperBound(): number | undefined {
+    return undefined;
+  }
+
+  get isArray(): boolean | undefined {
+    return undefined;
+  }
```

Comment:
Kinda confused by these methods being on EnumIdlNode... I would imagine that these properties would only exist on nodes that correspond to fields?

Reply:
removed